### PR TITLE
Tolerate eth errors

### LINKED ISSUE #462: Tolerate EthClient failures

### PR DESCRIPTION
**Pull request by alari**
Fixes #462

### CODE REVIEW

File: ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  /**
+   * Safe conversion from rx Flowable to fs2 Stream: rx is initiated again in case of any error.
+   *
+   * @param flowable RX Flowable
+   * @tparam T Value type
+   */
+  implicit class FlowableToStreamOps[T](flowable: Flowable[T]) extends LazyLogging {
+
+    def toStreamRetrying[F[_]: ConcurrentEffect: Timer](
```
Comment (folex) [inline ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala]: Maybe call it `retryUntilSuccess`? The same as method for retrying `F`.

File: ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      onErrorRetryAfter: FiniteDuration = EthRetryPolicy.Default.delayPeriod
```
Comment (folex) [inline ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala]: Why are you using scalar value here? Let's use `EthRetryPolicy`, it has nice back-off logic.

File: node/src/main/scala/fluence/node/eth/FluenceContract.scala [hunk 3]
```diff
@@ -156,10 +161,10 @@ class FluenceContract(private[eth] val ethClient: EthClient, private[eth] val co
    * @tparam F ConcurrentEffect to convert Observable into fs2.Stream
    * @return Possibly infinite stream of AppDeleted events
    */
-  private[eth] def getAppDeleted[F[_]: ConcurrentEffect]: fs2.Stream[F, Uint256] =
+  private[eth] def getAppDeleted[F[_]: ConcurrentEffect: Timer]: fs2.Stream[F, Uint256] =
     fs2.Stream
       .eval(eventFilter[F](APPDELETED_EVENT))
-      .flatMap(filter ⇒ contract.appDeletedEventFlowable(filter).toStream[F])
+      .flatMap(filter ⇒ contract.appDeletedEventFlowable(filter).toStreamRetrying[F](1.second))
```
Comment (folex) [inline node/src/main/scala/fluence/node/eth/FluenceContract.scala]: Why hardcode timeout?

File: node/src/main/scala/fluence/node/eth/FluenceContract.scala [hunk 4]
```diff
@@ -168,10 +173,10 @@ class FluenceContract(private[eth] val ethClient: EthClient, private[eth] val co
    * @tparam F ConcurrentEffect to convert Observable into fs2.Stream
    * @return Possibly infinite stream of NodeDeleted events
    */
-  private[eth] def getNodeDeleted[F[_]: ConcurrentEffect]: fs2.Stream[F, Bytes32] =
+  private[eth] def getNodeDeleted[F[_]: ConcurrentEffect: Timer]: fs2.Stream[F, Bytes32] =
     fs2.Stream
       .eval(eventFilter[F](NODEDELETED_EVENT))
-      .flatMap(filter ⇒ contract.nodeDeletedEventFlowable(filter).toStream[F]())
+      .flatMap(filter ⇒ contract.nodeDeletedEventFlowable(filter).toStreamRetrying[F](1.second))
```
Comment (folex) [inline node/src/main/scala/fluence/node/eth/FluenceContract.scala]: hardcoded timeout

**Comment by folex** [comment]
Also, other modules, e.g., `externalstorage`, could benefit from this code. Is it possible to make it available to other modules?

File: ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala (hunk 1 above)
Comment (alari) [reply ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala]: It doesn't stop retrying, so there's no `until`. It makes a stream that performs retrying under the hood. So meaning is different, I don't think names should be the same.

File: ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala (hunk 2 above)
Comment (alari) [reply ethclient/src/main/scala/fluence/ethclient/syntax/syntax.scala]: There's no straightforward back-off algorithm for streams. We don't know when the request has failed: before the first message, after N messages received? Was it the first time in a row when stream failed?

**Comment by alari** [comment]
Let's do code reuse for `externalstorage` in a separate PR.

**Comment by folex** [comment]
GitHub is lagging and not showing commits here. Closing for #469

**Pull request closed without merging**